polkadot/runtime/common/src/integration_tests.rs

```diff
@@ -1698,3 +1703,136 @@ fn cant_bid_on_existing_lease_periods() {
+
+		// Start a new auction in the future
+		let duration = 99u32;
+		let lease_period_index_start = 4u32;
+		assert_ok!(Auctions::new_auction(
+			RuntimeOrigin::root(),
+			duration,
+			lease_period_index_start
+		));
+
+		// 2 sessions later 2000 is an on-demand parachain
+		run_to_session(START_SESSION_INDEX + 2);
+		assert_eq!(Paras::lifecycle(ParaId::from(2000)), Some(ParaLifecycle::Parathread));
+
+		// Open a crowdloan for Para 1 for slots 0-3
+		assert_ok!(Crowdloan::create(
+			signed(1),
+			ParaId::from(2000),
+			1_000_000,                    // Cap
+			lease_period_index_start + 0, // First Slot
+			lease_period_index_start + 3, // Last Slot
+			200,                          // Block End
+			None,
+		));
+		let fund = Crowdloan::funds(ParaId::from(2000)).unwrap();
+		let crowdloan_account = Crowdloan::fund_account_id(fund.fund_index);
+
+		// Bunch of contributions
+		let mut total = 0;
+		for i in 10..20 {
+			Balances::make_free_balance_be(&account_id(i), 1_000_000_000);
+			assert_ok!(Crowdloan::contribute(signed(i), ParaId::from(2000), 900 - i, None));
+			total += 900 - i;
+		}
+		// Go to end of auction where everyone won their slots
+		run_to_block(200);
+
+		// deposit for validation code and crowdloan submission.
+		assert_eq!(Balances::reserved_balance(&account_id(1)), (500 + 10 * 2 * 1) + 100);
+		// crowdloan fund is reserved
+		assert_eq!(Balances::reserved_balance(&crowdloan_account), total);
+		// Crowdloan is appropriately set
+		assert!(Crowdloan::funds(ParaId::from(2000)).is_some());
+
+		// New leases will start on block 400
+		let lease_start_block = 400;
+		run_to_block(lease_start_block);
+
+		// Slots are won by Para 1
+		assert!(!Slots::lease(ParaId::from(2000)).is_empty());
+
+		// 2 sessions later it is a parachain
+		run_to_block(lease_start_block + 20);
+		assert_eq!(Paras::lifecycle(ParaId::from(2000)), Some(ParaLifecycle::Parachain));
+
+		// lease is active and not near end
+		assert_noop!(
+			Slots::early_lease_refund(signed(1), ParaId::from(2000)),
+			SlotsError::<Test>::NotAllowed
+		);
+
+		// Cant dissolve yet
+		assert_noop!(
+			Crowdloan::dissolve(signed(1), ParaId::from(2000)),
+			CrowdloanError::<Test>::NotReadyToDissolve
+		);
+
+		let lease_end_block = 8 * LeasePeriod::get() + LeaseOffset::get();
+		let early_refund_start_block = lease_end_block - EarliestRefundPeriod::get();
+		// Go in future when lease is about to end
+		run_to_block(early_refund_start_block);
+
+		// try early refund
+		assert_ok!(Slots::early_lease_refund(signed(1), ParaId::from(2000)));
+
+		// Withdraw of contributions works
+		assert_eq!(Balances::free_balance(&crowdloan_account), total);
+		for i in 10..20 {
+			assert_ok!(Crowdloan::withdraw(signed(i), account_id(i), ParaId::from(2000)));
+		}
+
+		assert_eq!(Balances::free_balance(&crowdloan_account), 0);
+
+		// reserved balance before dissolve
+		assert_eq!(Balances::reserved_balance(&account_id(1)), (500 + 10 * 2 * 1) + 100);
```

Comment:
nit: magic numbers

Reply:
Good point I will make it more descriptive